Added EmailChangedEvent

toxuin opened the pull request:
I've needed such events for integration with 3rd party services that use DB layout not really compatible with AuthMe.
My use case is that for every email change, this 3rd party service needs to send off it's own activation email which it stores in own DB.
If AuthMe would have such event, I could write a small tiny plugin that would talk to such 3rd party by, for example, REST API. This way, I could keep emails in sync.

Tests included.

Please let me know what you think.

games647 commented on src/main/java/fr/xephi/authme/events/EmailChangedEvent.java:
```diff
@@ -0,0 +1,54 @@
+        return player;
+    }
+
+    public String getOldEmail() {
```
Please document nullable values (including when/why) like here if the user adds a new email address.

games647 commented on src/main/java/fr/xephi/authme/process/email/AsyncAddEmail.java:
```diff
@@ -57,14 +62,19 @@ public void addEmail(Player player, String email) {
+                EmailChangedEvent event = bukkitService.createAndCallEvent(isAsync
+                    -> new EmailChangedEvent(player, null, email, isAsync));
+                if (event.isCancelled()) {
+                    sendFailedMessage(player);
```
In my opinion, this makes it hard to track down errors. `dataSource.updateEmail` will print an exception if an error occurrs, but if the event is cancelled you only get a very abstract error message.

My suggestion is to use a different error message for the console where you can see it's because the event was cancelled.

games647 commented on src/main/java/fr/xephi/authme/process/email/AsyncChangeEmail.java:
```diff
@@ -57,14 +62,21 @@ public void changeEmail(Player player, String oldEmail, String newEmail) {
+        EmailChangedEvent event = bukkitService.createAndCallEvent(isAsync
+            -> new EmailChangedEvent(player, oldEmail, newEmail, isAsync));
+        if (event.isCancelled()) {
+            service.send(player, MessageKey.ERROR);
```
Same here.

ljacqu commented:
This is quality work!

As @games647 already noted please distinguish both cases when logging. If changing an email fails because the event is canceled, the log level shouldn't be WARN—we reserve that for "unexpected" situations that should be investigated.

toxuin commented:
Hey, thanks for the kind words.

I've added some documentation for the new event. Now, depending on action performed (change\add), it would fire different messages off to user and console.

Well and since I'm at it, fixed up the TODOs in messages_ru.yml. Ура!

games647 commented on src/main/java/fr/xephi/authme/process/email/AsyncAddEmail.java:
```diff
@@ -57,6 +62,13 @@ public void addEmail(Player player, String email) {
+                EmailChangedEvent event = bukkitService.createAndCallEvent(isAsync
+                    -> new EmailChangedEvent(player, null, email, isAsync));
+                if (event.isCancelled()) {
+                    ConsoleLogger.warning("Could not add email to player '" + player + "' – event was cancelled");
```
Like @ljacqu said, warning is reserved for unexpected situation that should be investigated. Please change it to something else like info.

games647 commented on src/main/java/fr/xephi/authme/process/email/AsyncChangeEmail.java:
```diff
@@ -57,14 +63,22 @@ public void changeEmail(Player player, String oldEmail, String newEmail) {
+        EmailChangedEvent event = bukkitService.createAndCallEvent(isAsync
+            -> new EmailChangedEvent(player, oldEmail, newEmail, isAsync));
+        if (event.isCancelled()) {
+            ConsoleLogger.warning("Could not change email for player '" + player + "' – event was cancelled");
```
Same here

sgdc3 commented:
@toxuin Thank you very much for your contribution ;)
